### PR TITLE
WIP: add oneunit(x) for dimensionful version of one(x)

### PR DESCRIPTION
**Pull request by stevengj**
Following the discussion in #16116, this clarifies the documentation of `one(x)` to indicate that, since it is the multiplicative identity, it may not be the of the same type as `x` for types representing dimensionful quantities.  It defines the function `oneunit(x::T) = T(one(x))` for when you need the same type as `x` and `x` might be dimensionful.

I also went through Base and changed `one -> oneunit` where it seemed appropriate.  I didn't bother with usages of `one` in functions where a dimensionful quantity would not be meaningful, e.g. `exp(x)`, or where the type was otherwise constrained.

This change is not sufficient to make all of Base work with dimensionful quantities, but it is a start.  It would be good to have a `test/dimensionful.jl` test suite that began to exercise various base functions with dimensionful quantities (e.g. a minimalist version of SIUnits.jl).

### CODE REVIEW

**Comment by vchuravy** [comment]
I find `oneunit` quite a mouthful and I like `oneof` better. (as in `one of Type T`). But that is personal taste and I don't want to distract with syntax bike shedding.

+1 to introducing this concept.

**Comment by nalimilan** [comment]
Is a new function really needed since `T(1)` and `oftype(x, 1)` are quite short? (+1 for the changes in general though.)

File: base/number.jl
Comment (timholy) [inline base/number.jl]: I'm pretty sure that most people disliked the "shape" part, and that most think thought `one` on `Matrix{Float64}` should return a number (1.0, not `eye(n,n)`).

This PR doesn't have to do the deprecation, but I don't think we should put "shape" into the docstring.

File: base/number.jl
Comment (timholy) [inline base/number.jl]: "multiplicative identity" may not be clear to everyone, better to somewhere say "such that `one(x)*x == x*one(x) == x`".

File: base/number.jl
Comment (timholy) [inline base/number.jl]: You don't use this anywhere, and the name does not make it clear whether this would return a value or a type. Delete?

File: base/number.jl
Comment (timholy) [inline base/number.jl]: `oftype(one(x), -one(x))` might be safer than `oftype(one(x), -1)`.

**Review by timholy — COMMENTED** [review]
:+1:

**Comment by timholy** [comment]
@nalimilan, the voting (https://github.com/JuliaLang/julia/issues/16116#issuecomment-272712866) was almost evenly split on this issue.

**Comment by stevengj** [comment]
@nalimilan, `oftype(x, 1)` is  wrong here because it assumes that `1` is convertible to the type of `x`, which is not true for dimensionful quantities (e.g. `oftype(Dates.Day(3), 1)` fails), or quantities that are not numeric scalars but which still have an algebra (e.g. matrices).   In any case, I think `oneunit` is clearer to read, and it is also much easier to do a search-and-replace of `one` with `oneunit` to fix up usages.

File: base/number.jl
Comment (stevengj) [inline base/number.jl]: `x::Real` should make `-1` okay here.

File: base/number.jl
Comment (stevengj) [inline base/number.jl]: I didn't get the sense that there was any kind of consensus on whether `one(matrix)` should stop returning a matrix.  From an algebraic perspective, returning a matrix makes more sense.

Moreover, `oneunit(x)` only makes sense for matrices if `one` returns a matrix, since for `oneunit(x::T)` you have to be able to do `T(one(x))`, and this will fail if `T::Matrix` and `one(x)::Number`.

Comment (stevengj) [inline base/number.jl]: My perspective is that (a) `oneunit(x::T)` should be guaranteed to return something of type `T` and (b) `one` and `oneunit` should give the same result except for types representing dimensionful quantities (i.e. except for cases where the multiplicative identity *cannot* be the same type).

Comment (timholy) [inline base/number.jl]: >  It is also the right thing to do from the perspective of ^ and prod, both of which need the identity matrix for an empty product.

Not sure I see how, since `one(Matrix{Float64})` already doesn't work---you need at least one "sized" object to know what to return.

You're right that there are competing arguments. If you return a scalar, then `one` works in both the type-domain and the value-domain, whereas `oneunit` can't be defined. If you return a matrix, then you can't define `one` or `oneunit` in the type-domain.

I mostly think of these as "traits" and therefore the type-domain wins (not even a contest). But I recognize that this viewpoint is debatable.

Comment (stevengj) [inline base/number.jl]: @timholy, it works for `M^0`, since then it can return `one(M)`.

File: base/array.jl
Comment (tkelman) [inline base/array.jl]: shouldn't you be passing `x` to these?

**Comment by ajkeller34** [comment]
You could use the new definition of `eye` with a dimensionful type and wind up not creating the identity matrix for that type. In other words, if you had two matrices with `1.0m` along the diagonal and `0.0m` elsewhere, upon multiplication you would end up with a matrix having `1.0m^2` along the diagonal. The true identity matrix for `[1.0m 0.0m; 0.0m 1.0m]` would be `[1.0 0.0; 0.0 1.0]`.

It's not clear to me that this is the right way to go, but if this change is kept, perhaps we could reword the docstring for `eye` to something in the spirit of, but less clumsy than: "Constructs a matrix with the same dimensions and type as A, where `eltype(A)(1)` is along the diagonal and `eltype(A)(0)` is elsewhere. For dimensionless numbers, this is an identity matrix, but not in general." An alternative choice would be "Constructs an identity matrix of the same dimensions as A (and of the same type, where possible)."

**Comment by stevengj** [comment]
@ajkeller34, if you call `eye(X)` on a dimensionful matrix `X`, it should return the correct (dimensionless) multiplicative identity for `X`.   If you call `eye(T, m,n)`, it seems to me that it should create a `Matrix{T}`, however.

**Comment by ajkeller34** [comment]
I agree with you, I think. To be clear though, are you saying you are fine with `eye(T,m,n)` not returning an identity matrix for dimensionful `T`? Or would you rather `eye(T,m,n)` fail for dimensionful `T`, so that `eye` will only ever return an identity matrix?

**Comment by ajkeller34** [comment]
Ah, never mind, didn't see your updated comment.

**Comment by stevengj** [comment]
@timholy, any further comments?

**Review by timholy — APPROVED** [review]
LGTM

**Comment by stevengj** [comment]
Rebased.  Good to merge when green?

File: doc/src/manual/style-guide.md
Comment (tkelman) [inline doc/src/manual/style-guide.md]: shame to mess with the comment spacing in the style guide

**Comment by timholy** [comment]
Thanks for tackling this, @stevengj.

**Comment by timholy** [comment]
Oh, maybe I'll let you merge. I'd prefer to squash this down, "whoops" is not the most informative commit message :smile:.

**Comment by stevengj** [comment]
@timholy, the "squash and merge" button lets you edit the commit message.  I usually trim out all the extraneous commit messages at that point.

**Comment by timholy** [comment]
I'm well aware, I just didn't want to take liberties in case there was a reason to have separate commits.

**Pull request merged**
